p2panda-sync/src/log_sync.rs

```diff
@@ -149,76 +166,87 @@ where
-                    for log_id in log_ids {
-                        let local_log_heights =
-                            self.store.get_log_heights(&log_id).await.map_err(|err| {
-                                SyncError::Critical(format!(
-                                    "can't retreive log heights from store, {err}"
-                                ))
-                            })?;
-
-                        for (public_key, seq_num) in local_log_heights {
-                            let Some(remote_log_heights) = remote_log_heights_map.get(&log_id)
-                            else {
-                                // The remote peer didn't request logs with this id
-                                continue;
-                            };
-
-                            for (remote_pub_key, remote_seq_num) in remote_log_heights.iter() {
-                                // Compare log heights sent by the remote with our local logs, if
-                                // our logs are more advanced calculate and send operations the
-                                // remote is missing
-                                if *remote_pub_key == public_key && *remote_seq_num < seq_num {
-                                    let messages = remote_needs(
-                                        &self.store,
-                                        &log_id,
-                                        &public_key,
-                                        *remote_seq_num + 1,
-                                    )
-                                    .await?;
-                                    sink.send_all(&mut stream::iter(messages.into_iter().map(Ok)))
+                    for (public_key, log_ids) in logs {
+                        let Some(remote_log_heights) = remote_log_heights_map.get(&public_key)
+                        else {
+                            for log_id in log_ids {
+                                let messages =
+                                    remote_needs::<T, L, E>(&self.store, &log_id, &public_key, 0)
                                         .await?;
-                                };
+                                sink.send_all(&mut stream::iter(messages.into_iter().map(Ok)))
+                                    .await?;
                             }
+                            continue;
+                        };
+
+                        for log_id in log_ids {
+                            let latest_operation = self
+                                .store
+                                .latest_operation(&public_key, &log_id)
+                                .await
+                                .map_err(|err| {
+                                    SyncError::Critical(format!(
+                                        "can't retreive log heights from store, {err}"
+                                    ))
+                                })?;
```

Comment:
We probably want a method for this in the store (`latest_operations(log_ids: Vec<_>)`) for optimal store queries, just noting that here and maybe we can make a ticket?

Reply:
https://github.com/p2panda/p2panda/issues/579